src/rpc/mining.cpp

```diff
@@ -100,6 +100,88 @@ static UniValue getnetworkhashps(const JSONRPCRequest& request)
+    CChainParams chainparams(Params());
+
+    if (!preselected_txs) {
+        std::unique_ptr<CBlockTemplate> pblocktemplate(BlockAssembler(mempool, chainparams).CreateNewBlock(coinbase_script));
+        if (!pblocktemplate.get())
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Couldn't create new block");
+        block = pblocktemplate->block;
+        {
+            LOCK(cs_main);
+            IncrementExtraNonce(&block, ::ChainActive().Tip(), extra_nonce);
+        }
+    } else {
+
+        CBlockIndex* previous_index;
+        {
+            LOCK(cs_main);
+            previous_index = ::ChainActive().Tip();
+        }
+        CHECK_NONFATAL(previous_index != nullptr);
+
+        const int height = previous_index->nHeight + 1;
+
+        // Create coinbase transaction.
+        CMutableTransaction coinbase_tx;
+        coinbase_tx.vin.resize(1);
+        coinbase_tx.vin[0].prevout.SetNull();
+        coinbase_tx.vout.resize(1);
+        coinbase_tx.vout[0].scriptPubKey = coinbase_script;
+        coinbase_tx.vout[0].nValue = GetBlockSubsidy(height, chainparams.GetConsensus());
+        coinbase_tx.vin[0].scriptSig = CScript() << height << OP_0;
+        block.vtx.push_back(MakeTransactionRef(std::move(coinbase_tx)));
+
+        // Add transactions
+        block.vtx.insert(block.vtx.end(), preselected_txs->begin(), preselected_txs->end());
+
+        block.nVersion = ComputeBlockVersion(previous_index, chainparams.GetConsensus());
+        if (chainparams.MineBlocksOnDemand())
+            block.nVersion = gArgs.GetArg("-blockversion", block.nVersion);
+
+        // Fill in header
+        block.hashPrevBlock = previous_index->GetBlockHash();
+        block.nTime = GetAdjustedTime();
+        UpdateTime(&block, chainparams.GetConsensus(), previous_index);
+        block.nBits = GetNextWorkRequired(previous_index, &block, chainparams.GetConsensus());
+        block.nNonce = 0;
+
+        GenerateCoinbaseCommitment(block, previous_index, chainparams.GetConsensus());
+
+        {
+            LOCK(cs_main);
+            IncrementExtraNonce(&block, ::ChainActive().Tip(), extra_nonce);
+
+            BlockValidationState state;
+            if (!TestBlockValidity(state, chainparams, block, previous_index, false, false)) {
+                throw JSONRPCError(RPC_VERIFY_ERROR, strprintf("TestBlockValidity failed: %s", state.ToString()));
+            }
+        }
+    }
+
+    while (max_tries > 0 && block.nNonce < std::numeric_limits<uint32_t>::max() && !CheckProofOfWork(block.GetHash(), block.nBits, chainparams.GetConsensus()) && !ShutdownRequested()) {
+        ++block.nNonce;
+        --max_tries;
+    }
+    if (max_tries == 0 || ShutdownRequested()) {
+        return false;
+    }
+    if (block.nNonce == std::numeric_limits<uint32_t>::max()) {
+        return true;
+    }
+    std::shared_ptr<const CBlock> shared_pblock = std::make_shared<const CBlock>(block);
+    if (!ProcessNewBlock(chainparams, shared_pblock, true, nullptr))
```

Comment:
comment for followup only: Allow a bypass of validity check for building block tests via RPC